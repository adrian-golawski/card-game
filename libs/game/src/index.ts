export * from './lib/game.module';
export * from './lib/game-container/game-container.component';
