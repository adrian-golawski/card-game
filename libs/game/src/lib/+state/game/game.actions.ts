import { createAction, props } from '@ngrx/store';

export const startNewGame = createAction('[Game] Start New Game');
export const startNewGameRequest = createAction(
  '[Game] Start New Game Request'
);

export const startNewGameSuccess = createAction(
  '[Game] Start New Game Success'
);
export const startNewGameFailure = createAction(
  '[Game] Start New Game Failure'
);

export const betLower = createAction(
  '[Game] Bet lower',
  props<{ lower: boolean }>()
);
