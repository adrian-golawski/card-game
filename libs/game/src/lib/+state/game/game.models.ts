export interface GameEntity {
  id: number;
}
