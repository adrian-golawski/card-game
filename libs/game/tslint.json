{
  "extends": "../../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "cardGame", "camelCase"],
    "component-selector": [true, "element", "card-game", "kebab-case"]
  },
  "linterOptions": {
    "exclude": ["!**/*"]
  }
}
