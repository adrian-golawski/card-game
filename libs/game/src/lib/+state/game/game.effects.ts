import { Injectable } from '@angular/core';
import * as CardActions from '@card-game/cards';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import * as GameActions from './game.actions';

@Injectable()
export class GameEffects {
  constructor(private readonly actions$: Actions) {}

  public startNewGame$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(GameActions.startNewGame),
      map(CardActions.getNewDeck)
    )
  );
}
