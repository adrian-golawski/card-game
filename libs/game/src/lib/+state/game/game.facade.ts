import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';

import { Observable } from 'rxjs';

import * as GameActions from './game.actions';
import * as fromGame from './game.reducer';
import * as GameSelectors from './game.selectors';

@Injectable()
export class GameFacade {
  score$: Observable<number> = this.store.pipe(select(GameSelectors.getScore));
  gameInProgress$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.gameInProgress)
  );
  roundsLeft$: Observable<number> = this.store.pipe(
    select(GameSelectors.roundsLeft)
  );
  betGiven$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.betGiven)
  );
  betLower$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.betLower)
  );

  constructor(private readonly store: Store<fromGame.GamePartialState>) {}

  startNewGame(): void {
    this.store.dispatch(GameActions.startNewGame());
  }

  betLower(lower: boolean): void {
    this.store.dispatch(GameActions.betLower({ lower }));
  }
}
