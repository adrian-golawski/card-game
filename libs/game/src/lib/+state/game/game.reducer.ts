import { Action, createReducer, on } from '@ngrx/store';

import * as GameActions from './game.actions';
export const GAME_FEATURE_KEY = 'game';

export interface State {
  score: number;
  gameActive?: boolean;
  roundsLeft?: number;
}

export interface GamePartialState {
  readonly [GAME_FEATURE_KEY]: State;
}

export const initialState: State = {
  score: 0,
};

const gamesReducer = createReducer(
  initialState,
  on(GameActions.startNewGameSuccess, (state) => ({
    ...state,
    score: 0,
    gameActive: true,
    roundsLeft: 30,
  }))
);

export function reducer(state: State, action: Action): State {
  return gamesReducer(state, action);
}
