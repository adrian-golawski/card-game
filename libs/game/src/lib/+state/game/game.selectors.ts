import { createFeatureSelector, createSelector } from '@ngrx/store';

import { GamePartialState, GAME_FEATURE_KEY, State } from './game.reducer';

export const getGameState = createFeatureSelector<GamePartialState, State>(
  GAME_FEATURE_KEY
);

export const getScore = createSelector(
  getGameState,
  (state: State) => state.score
);

export const gameInProgress = createSelector(
  getGameState,
  (state: State) => state.gameActive
);

export const roundsLeft = createSelector(
  getGameState,
  (state: State) => state.roundsLeft
);

export const betGiven = createSelector(
  getGameState,
  (state: State) => state.betGiven
);

export const betLower = createSelector(
  getGameState,
  (state: State) => state.betLower
);
