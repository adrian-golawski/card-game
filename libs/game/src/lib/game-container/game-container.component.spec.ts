import { GameContainerComponent, GameModule } from '@card-game/game';

import { Shallow } from 'shallow-render';

describe('GameContainerComponent', () => {
  let shallow: Shallow<GameContainerComponent>;

  beforeEach(() => {
    shallow = new Shallow(GameContainerComponent, GameModule);
  });

  it('should match the snapshot', async () => {
    const { fixture } = await shallow.render();

    expect(fixture).toMatchSnapshot();
  });
});
