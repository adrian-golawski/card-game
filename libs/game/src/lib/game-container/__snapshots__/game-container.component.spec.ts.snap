// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameContainerComponent should match the snapshot with existing played cards 1`] = `
<ng-component>
  <card-game-game-container>
    <div>
      Rounds left: 30
    </div>
    <div>
      Played Cards: [H2]
    </div>
    <button>
      Draw new Card
    </button>
    <div>
      <button>
         Bet lower 
      </button>
      <button>
         Bet higher 
      </button>
    </div>
    <div>
      Score: 0
    </div>
  </card-game-game-container>
</ng-component>
`;
